### PR TITLE
PSTRESS-85 Add support for ALTER TABLE .. DISCARD TABLESPACE

### PR DESCRIPTION
**Pull request by mohitj1988**

### CODE REVIEW

**Review by rahulmalik87 — CHANGES_REQUESTED** [review]
If wediscarding the tablespace the table will become useless. 
so either we need to import the tablespace again.( you need to save ibd first somewhere )
or do it drop and recreated after discard

**Comment by mohitj1988** [comment]
> If we arediscarding the tablespace, the table will become useless.
> so either we need to import the tablespace again.( you need to save ibd first somewhere )
> or do it drop and recreated after discard

Makes sense. We can drop and create the table. Implemented it this way.
Until upstream bug https://jira.percona.com/browse/PS-7865 is not fixed, we can disable this feature

File: src/help.cpp [hunk 1]
```diff
@@ -367,6 +367,15 @@ void add_options() {
   opt->setSQL();
   opt->setDDL();
 
+  /*Discard tablespace */
+  opt = newOption(Option::INT, Option::ALTER_DISCARD_TABLESPACE,
+		  "alt-discard-tbs");
+  opt->help = "Alter table to discard file-per-tablespace";
+  /* Disabled by default, until https://jira.percona.com/browse/PS-7865 is fixed */
+  opt->setInt(0);
```
Comment (rahulmalik87) [inline src/help.cpp:375]: disable it only version lower than 8.0.26. Hope they will fix it 8.0.27

File: src/random_test.cpp [hunk 2]
```diff
@@ -89,13 +88,18 @@ int sum_of_all_options(Thd1 *thd) {
...
-  if (db_branch().compare("5.7") == 0) {
+  if (db_branch().substr(0,3).compare("5.7") == 0) {
     opt_int_set(ALTER_TABLESPACE_RENAME, 0);
     opt_int_set(RENAME_COLUMN, 0);
     opt_int_set(UNDO_SQL, 0);
     opt_int_set(ALTER_REDO_LOGGING, 0);
   }
 
+  /* Disabling until Bug: https://jira.percona.com/browse/PS-7865 is fixed by upstream */
+  if (db_branch().substr(0,6).compare("8.0.26") == 0) {
```
Comment (rahulmalik87) [inline src/random_test.cpp:99]: We need to disable it for lower version as well. not just only for 8.0.26.

File: src/random_test.cpp [hunk 3]
```diff
@@ -54,7 +54,6 @@ static std::string get_result(std::string sql, Thd1 *thd) {
 /* return the db branch */
 static std::string db_branch() {
```
Comment (rahulmalik87) [inline src/random_test.cpp:55]: can you modify to use method like 
/*
    Get version number for server in a form easy to test on

    SYNOPSIS
      mysql_get_server_version()
      mysql   Connection

    EXAMPLE
      4.1.0-alfa ->  40100

    NOTES
      We will ensure that a newer server always has a bigger number.

    RETURN
     Signed number > 323000
     Zero if there is no connection
  */

* find the pxb base version */
  static unsigned long pxb_base_version() {
    std::string pxb_base = MYSQL_SERVER_VERSION;
    unsigned long major = 0, minor = 0, version = 0;
    std::size_t major_p = pxb_base.find(".");
    if (major_p != std::string::npos) major = stoi(pxb_base.substr(0, major_p));

    std::size_t minor_p = pxb_base.find(".", major_p + 1);
    if (minor_p != std::string::npos)
      minor = stoi(pxb_base.substr(major_p + 1, minor_p - major_p));

    std::size_t version_p = pxb_base.find(".", minor_p + 1);
    if (version_p != std::string::npos)
      version = stoi(pxb_base.substr(minor_p + 1, version_p - minor_p));
    else
      version = stoi(pxb_base.substr(minor_p + 1));
    return major * 10000 + minor * 100 + version;
  }

Comment (rahulmalik87) [reply src/random_test.cpp:55]: so with this. we can do something like this if version < 80000. then disable some feature for 5.7.  also and version < 80026

Comment (mohitj1988) [reply src/random_test.cpp]: Fixed as suggested

File: src/random_test.cpp (hunk 2 above)
Comment (mohitj1988) [reply src/random_test.cpp]: Fixed

File: src/help.cpp (hunk 1 above)
Comment (mohitj1988) [reply src/help.cpp]: Fixed

File: src/random_test.cpp [hunk 4]
```diff
@@ -51,11 +51,27 @@ static std::string get_result(std::string sql, Thd1 *thd) {
   return result;
 }
 
-/* return the db branch */
-static std::string db_branch() {
-  std::string branch = mysql_get_client_info();
-  branch = branch.substr(0, 3); // 8.0 or 5.7
-  return branch;
+/* return PS version in number format
```
Comment (rahulmalik87) [inline src/random_test.cpp:54]: replace ps with ms.

Comment (mohitj1988) [reply src/random_test.cpp:54]: We are using it with Percona software and also it is buit within Percona. Should we not use PS?

File: src/random_test.cpp [hunk 5]
```diff
@@ -51,11 +51,27 @@ static std::string get_result(std::string sql, Thd1 *thd) {
...
+  std::size_t major_p = ps_base.find(".");
+  if(major_p != std::string::npos) major = stoi(ps_base.substr(0, major_p));
+
+  std::size_t minor_p = ps_base.find(".", major_p + 1);
+  if (minor_p != std::string::npos)
+    minor = stoi(ps_base.substr(major_p + 1, minor_p - major_p));
+
+  std::size_t version_p = ps_base.find(".", minor_p + 1);
+  if (version_p != std::string::npos)
+    version = stoi(ps_base.substr(minor_p + 1, version_p - minor_p));
```
Comment (rahulmalik87) [inline src/random_test.cpp:70]: check if it should be version or version_p

File: src/random_test.cpp (hunk 4 above)
Comment (mohitj1988) [reply src/random_test.cpp]: Replaced "ps" with the word "server"

File: src/random_test.cpp (hunk 5 above)
Comment (mohitj1988) [reply src/random_test.cpp:70]: It should be "version". The existing code is fine.

**Review by rahulmalik87 — APPROVED** [review]
LGTM

**Pull request merged**